File: example/src/main/java/com/github/simonpercic/oklogexample/ui/base/DaggerActivity.java
package com.github.simonpercic.oklogexample.ui.base;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;

import com.github.simonpercic.oklogexample.App;
import com.github.simonpercic.oklogexample.AppComponent;

/**
 * Dagger base Activity.
 *
 * @author Simon Percic <a href="https://github.com/simonpercic">https://github.com/simonpercic</a>
 */
public abstract class DaggerActivity<T> extends AppCompatActivity {

    private T component;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        component = inject();
    }

    /**
     * Inject Dagger components.
     *
     * @return injected Dagger component.
     */
    protected abstract T inject();

    public T getComponent() {
        return component;
    }

    protected AppComponent getAppComponent() {
        return App.get(this).getComponent();
    }
}

File: example/src/main/kotlin/com/github/simonpercic/oklogexample/ui/base/DaggerActivity.kt
package com.github.simonpercic.oklogexample.ui.base

import android.os.Bundle
import android.support.v7.app.AppCompatActivity

import com.github.simonpercic.oklogexample.App
import com.github.simonpercic.oklogexample.AppComponent

/**
 * Dagger base Activity.

 * @author Simon Percic [https://github.com/simonpercic](https://github.com/simonpercic)
 */
abstract class DaggerActivity<T : Any> : AppCompatActivity() {

    protected lateinit var component: T
        private set

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        component = inject()
    }

    /**
     * Inject Dagger components.

     * @return injected Dagger component.
     */
    protected abstract fun inject(): T

    protected val appComponent: AppComponent
        get() = App.get(this).component
}
